File: lib/email.js
/**
 * 邮件发送 — 使用 Resend
 * ---------------------------------------------------------------------------
 * 所有配置项(RESEND_API_KEY / RESEND_FROM / SITE_NAME)都从 settings 表读。
 * 未配置时进入开发模式:不真实发送,code 原样通过返回值告诉前端显示。
 * 不再向 data/emails/ 写本地副本(避免单进程持续 IO 与潜在磁盘堆积)。
 *
 * 模板设计:
 *   • 整体走浅色 —— 米白(#faf8f5)底 + 奶油白(#fff)卡片,不再使用深色顶栏。
 *     邮件客户端的深色模式多数会自动反色,我们用浅色基调,收件方不管是白
 *     还是深色主题,观感都稳定,不会出现"深上加深"的糊色。
 *   • 图标与官网顶栏的 BrandIcon(rings 变体 + jade 配色)一致 —— 同心环 +
 *     中心点,翠绿描边。SVG 内联,邮件客户端不会去外链请求,也就不会因为
 *     图片拦截策略而裂图。viewBox / 尺寸与 BrandIcon 完全同步,便于日后改
 *     样式时只要同步这两处(BrandIcon.js 与此文件的 brandMarkSvg)。
 * ---------------------------------------------------------------------------
 */
import { getSetting } from './settings.js';

// 品牌色 —— 与 components/ui/BrandIcon.js 的 jade 调色板保持一致
const BRAND_STROKE = '#4fa87b';
const BRAND_ACCENT = '#2f7c5a';
const BRAND_BG     = '#eff7f2';
const BRAND_BORDER = '#cfe6d8';

/**
 * 和 BrandIcon 的 rings 变体严格同步(同心环 + 中心点)。
 * 如果将来在 BrandIcon.js 里切换默认图案或改色,记得也把这里同步过去 ——
 * 这里是邮件里的品牌图标,无法动态引用客户端组件。
 */
function brandMarkSvg(size = 38) {
  return `<svg viewBox="0 0 48 48" width="${size}" height="${size}" xmlns="http://www.w3.org/2000/svg" role="img" aria-label="栖枢" style="display:block;">
  <rect x="0.5" y="0.5" width="47" height="47" rx="11" ry="11" fill="${BRAND_BG}" stroke="${BRAND_BORDER}" stroke-width="1"/>
  <circle cx="24" cy="24" r="13" fill="none" stroke="${BRAND_STROKE}" stroke-width="2.25"/>
  <circle cx="24" cy="24" r="7"  fill="none" stroke="${BRAND_STROKE}" stroke-width="1.4" opacity="0.75"/>
  <circle cx="24" cy="24" r="2.4" fill="${BRAND_ACCENT}"/>
</svg>`;
}

function codeEmailHtml(siteName, title, body) {
  // 外层 table 兼容性更好(Outlook 对 div/flex 支持糟糕,table 稳)。
  // 顶栏使用奶油白 + 浅米色下边框,与官网顶栏风格一致(都是浅色+淡边)。
  return `<!DOCTYPE html>
<html>
<head><meta charset="utf-8"></head>
<body style="margin:0;padding:0;background:#faf8f5;font-family:'PingFang SC','Hiragino Sans GB',system-ui,-apple-system,Segoe UI,sans-serif;color:#1a1612;">
  <table width="100%" cellpadding="0" cellspacing="0" border="0" style="background:#faf8f5;padding:40px 16px;">
    <tr>
      <td align="center">
        <table width="520" cellpadding="0" cellspacing="0" border="0" style="max-width:520px;width:100%;background:#ffffff;border-radius:12px;border:1px solid #e5dfd9;overflow:hidden;">
          <tr>
            <td style="background:#ffffff;padding:22px 32px;border-bottom:1px solid #efe9e2;">
              <table cellpadding="0" cellspacing="0" border="0">
                <tr>
                  <td style="vertical-align:middle;padding-right:12px;">${brandMarkSvg(38)}</td>
                  <td style="vertical-align:middle;">
                    <div style="font-size:1.1rem;font-weight:600;color:#1a1612;letter-spacing:-0.01em;line-height:1.1;">${siteName}</div>
                    <div style="font-size:0.68rem;color:#9e9087;letter-spacing:0.22em;margin-top:3px;text-transform:uppercase;">Qi · Shu</div>
                  </td>
                </tr>
              </table>
            </td>
          </tr>
          <tr>
            <td style="padding:28px 32px 8px;">
              <h2 style="margin:0 0 12px;color:#1a1612;font-size:1.2rem;font-weight:600;">${title}</h2>
              <p style="color:#6b5f57;font-size:0.92rem;line-height:1.75;margin:0 0 20px;">${body}</p>
            </td>
          </tr>
          <tr>
            <td style="padding:4px 32px 26px;color:#9e9087;font-size:0.78rem;line-height:1.6;">
              此邮件由 ${siteName} 平台自动发送,请勿回复。
            </td>
          </tr>
        </table>
      </td>
    </tr>
  </table>
</body>
</html>`;
}

async function sendViaResend({ apiKey, from, siteName }, to, subject, html) {
  const { Resend } = await import('resend');
  const resend = new Resend(apiKey);
  const { error } = await resend.emails.send({
    from: `${siteName} <${from}>`,
    to,
    subject,
    html,
  });
  if (error) throw new Error(`Resend error: ${error.message}`);
}

function loadEmailConfig() {
  return {
    siteName: getSetting('SITE_NAME') || '栖枢',
    apiKey: getSetting('RESEND_API_KEY') || '',
    from: getSetting('RESEND_FROM') || 'noreply@qishu.local',
  };
}

/**
 * 发送邮件的公共入口。
 * ---------------------------------------------------------------------------
 * 历史问题(#5):生产环境下,若 RESEND_API_KEY 未配置,原实现会走"开发模式":
 *   - 把验证码 console.log 到 stdout(日志里泄露敏感码)
 *   - 返回 { code },让上层 API 假装成功
 *   - 上层 API 又靠 NODE_ENV !== 'production' 过滤不把 code 回给前端
 *   - 最终结果:前端看到"邮件已发送",但用户邮箱永远收不到
 * 这条链路在生产上就是一个"注册看似成功,但用户完成不了"的静默黑洞。
 *
 * 修复策略:
 *   - NODE_ENV === 'production' 且 apiKey 为空 → 抛错,由上层 500,不走 dev 回显
 *   - 否则(开发态) → 保留旧行为:console.log + 返回 { code } 供前端回显
 * 上层 API (register / forgot-password / change-password 等)继续靠
 * NODE_ENV 决定是否把 devCode 回给前端,两层策略互相独立、叠加安全。
 * ---------------------------------------------------------------------------
 */
async function sendCodeEmail({ to, subject, html, code }) {
  const cfg = loadEmailConfig();

  if (!cfg.apiKey) {
    if (process.env.NODE_ENV === 'production') {
      // 在生产环境下,没有 apiKey = 真的发不出邮件。必须硬失败,不能静默成功。
      // 抛的错上游会被 catch,对外响应 500;运营能在日志里立刻看到这条,并去
      // /admin/settings 把 RESEND_API_KEY 填上。比用户反馈"我收不到验证码"
      // 早几天发现问题。
      throw new Error('邮件服务未配置(RESEND_API_KEY 为空),无法发送邮件');
    }
    // 开发态:把 code 打到 stdout 并返回给上层,让前端在 NODE_ENV!==production
    // 时能回显出来,便于本地调试。
    console.log(`[DEV EMAIL] To: ${to} | Subject: ${subject} | Code: ${code}`);
    return { code };
  }

  await sendViaResend(cfg, to, subject, html);
  return {};
}

export async function sendVerificationCode(email, code) {
  const cfg = loadEmailConfig();
  const subject = `${cfg.siteName} 邮箱验证码`;
  const html = codeEmailHtml(
    cfg.siteName,
    '邮箱验证码',
    `您正在注册 ${cfg.siteName} 账号,验证码为:<br>
     <span style="font-size:2rem;font-weight:700;color:#1a1612;letter-spacing:0.2em;">${code}</span><br>
     验证码有效期 <strong>30 分钟</strong>,请勿泄露给他人。`
  );
  return sendCodeEmail({ to: email, subject, html, code });
}

export async function sendPasswordResetCode(email, code) {
  const cfg = loadEmailConfig();
  const subject = `${cfg.siteName} 密码重置验证码`;
  const html = codeEmailHtml(
    cfg.siteName,
    '密码重置验证码',
    `您正在重置 ${cfg.siteName} 账号密码,验证码为:<br>
     <span style="font-size:2rem;font-weight:700;color:#1a1612;letter-spacing:0.2em;">${code}</span><br>
     验证码有效期 <strong>30 分钟</strong>。如非本人操作,请忽略此邮件。`
  );
  return sendCodeEmail({ to: email, subject, html, code });
}

File: internal/email/email.go
// Package email sends transactional mail via Resend's HTTP API.
//
// 安全策略修改 (2026-04):
//   - 移除"开发模式 stdout 回显验证码"的旧行为。任何环境下,如果
//     RESEND_API_KEY 未配置,Send 都返回 ErrNotConfigured,调用方
//     必须把这个错误转成对终端用户可见的 503 + 提示文案
//     "服务器未配置邮件发送服务,请联系管理员"。
//   - 验证码绝对不会被回显到 HTTP 响应体或服务器日志中。
//
// 模板美化 (2026-05):
//   - 验证码邮件改用柔绿玻璃风格,与产品主题一致;邮件顶部显示
//     "栖枢"产品图标(emerald 渐变方块 + 白色 sparkle),保证品牌识别。
package email

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"html"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

// ErrNotConfigured 表示当前没有可用的邮件发送通道(RESEND_API_KEY 留空)。
// 所有依赖发送邮件验证码的 handler 收到这个错误时,必须直接告诉用户:
// 邮件服务尚未配置,请联系管理员;不要静默吞掉。
var ErrNotConfigured = errors.New("email sender not configured")

// Sender 对外暴露的接口。Send 调用线程安全,Reload 也线程安全。
type Sender struct {
	mu       sync.RWMutex
	apiKey   string
	from     string
	httpDone *http.Client
}

func New(apiKey, from string) *Sender {
	return &Sender{
		apiKey:   strings.TrimSpace(apiKey),
		from:     strings.TrimSpace(from),
		httpDone: &http.Client{Timeout: 10 * time.Second},
	}
}

// Reload 切换 API Key/From,allowing admin 修改 settings 立即生效。
func (s *Sender) Reload(apiKey, from string) {
	s.mu.Lock()
	s.apiKey = strings.TrimSpace(apiKey)
	s.from = strings.TrimSpace(from)
	s.mu.Unlock()
}

// Configured 报告当前是否已配置 API Key + From。
func (s *Sender) Configured() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.apiKey != "" && s.from != ""
}

// Send 发送邮件。如果 RESEND_API_KEY 未配置,直接返回 ErrNotConfigured —
// 不再回退到 stdout 打日志,也不再"静默成功"。
func (s *Sender) Send(ctx context.Context, to, subject, htmlBody, textBody string) error {
	s.mu.RLock()
	apiKey := s.apiKey
	from := s.from
	client := s.httpDone
	s.mu.RUnlock()

	if apiKey == "" {
		return ErrNotConfigured
	}
	if from == "" {
		return errors.New("RESEND_FROM 未设置,无法发件")
	}

	body := resendRequest{From: from, To: to, Subject: subject, HTML: htmlBody, Text: textBody}
	buf, err := json.Marshal(body)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		"https://api.resend.com/emails", bytes.NewReader(buf))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		errMsg, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return fmt.Errorf("resend %d: %s", resp.StatusCode, string(errMsg))
	}
	return nil
}

type resendRequest struct {
	From    string `json:"from"`
	To      string `json:"to"`
	Subject string `json:"subject"`
	HTML    string `json:"html"`
	Text    string `json:"text"`
}

// ─── Templates ────────────────────────────────────────────────────────────

// 验证码邮件模板 — 柔绿玻璃风格,与产品主题一致(详见 web/index.html)。
//
// 构成:
//  1. 顶部品牌图标(emerald 渐变圆角方块 + 白色 sparkle)+ 站点名
//  2. 白色卡片承载正文与验证码块
//  3. 验证码块用极浅薄荷底,大号 monospace 数字 + letter-spacing
//
// 邮件客户端兼容:
//   - 使用 <table> 布局而非 flex/grid,Outlook 桌面也能渲染
//   - 品牌图标的渐变在 Outlook 桌面下会被忽略,所以同时给 td 写
//     bgcolor + background-color fallback,最差情况也能显示纯品牌色
//   - 中心 sparkle 是 inline <svg>,Outlook 会忽略,但底色块仍在,
//     不会破坏整体视觉
func ComposeVerificationCode(siteName, action, code string, expiryMin int) (subject, htmlBody, textBody string) {
	if siteName == "" {
		siteName = "Qi Shu"
	}
	if action == "" {
		action = "完成账号验证"
	}
	subject = "[" + siteName + "] 验证码: " + code

	textBody = fmt.Sprintf(`%s

你正在%s%s。请在 %d 分钟内输入以下验证码:

  %s

如果你并未发起此操作,请忽略本邮件。`, siteName, siteName, action, expiryMin, code)

	site := html.EscapeString(siteName)
	act := html.EscapeString(action)
	c := html.EscapeString(code)
	htmlBody = `<!doctype html><html lang="zh-CN"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><title>` + site + ` 验证码</title></head>
<body style="margin:0;padding:0;background:#F4FBF6;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft Yahei',sans-serif;color:#1F3B2A">
<table role="presentation" width="100%" cellpadding="0" cellspacing="0" border="0" style="background:#F4FBF6">
  <tr><td align="center" style="padding:40px 16px">
    <table role="presentation" width="560" cellpadding="0" cellspacing="0" border="0" style="max-width:560px;width:100%">
      <tr><td align="center" style="padding-bottom:20px">` + brandIconHTML(56) + `</td></tr>
      <tr><td align="center" style="padding:0 0 28px 0">
        <div style="font-size:18px;font-weight:600;letter-spacing:0.5px;color:#0F4A33">` + site + `</div>
      </td></tr>
      <tr><td style="background:#FFFFFF;border:1px solid rgba(15,36,25,0.08);border-radius:18px;padding:36px 36px 32px 36px;box-shadow:0 12px 32px -16px rgba(15,36,25,0.10)">
        <div style="font-size:15px;line-height:1.7;color:#1F3B2A;margin:0 0 8px 0">你好,</div>
        <div style="font-size:15px;line-height:1.7;color:#1F3B2A;margin:0 0 22px 0">你正在 <strong style="color:#047857">` + site + `</strong> ` + act + `。请在 <strong>` + fmt.Sprintf("%d", expiryMin) + ` 分钟</strong>内输入以下验证码完成验证:</div>
        <div style="background:linear-gradient(180deg,#ECFDF5 0%,#F4FBF6 100%);background-color:#ECFDF5;border:1px solid rgba(16,185,129,0.25);border-radius:14px;padding:24px;text-align:center">
          <div style="font-family:'JetBrains Mono','SF Mono',Menlo,Consolas,monospace;font-size:34px;font-weight:600;letter-spacing:10px;color:#064E3B;line-height:1">` + c + `</div>
          <div style="font-size:11px;color:#6B7C73;margin-top:10px;letter-spacing:0.5px">VERIFICATION CODE</div>
        </div>
        <div style="font-size:13px;line-height:1.7;color:#6B7C73;margin:22px 0 0 0">为了你的账号安全,请勿向任何人转发此验证码。如果你并未发起此操作,可放心忽略本邮件,你的账号不会受到任何影响。</div>
      </td></tr>
      <tr><td align="center" style="padding:20px 8px 0 8px">
        <div style="font-size:12px;color:#94A39A;line-height:1.6">本邮件由系统自动发送,请勿直接回复。<br>© ` + site + `</div>
      </td></tr>
    </table>
  </td></tr>
</table>
</body></html>`
	return subject, htmlBody, textBody
}

// brandIconHTML 渲染品牌图标 — emerald 渐变圆角方块 + 中心白色 sparkle,
// 用 <table bgcolor=...><svg/></table> 双层兜底,Outlook 桌面忽略 SVG 时
// 仍能显示纯品牌色块。
func brandIconHTML(size int) string {
	radius := size / 4
	starSize := int(float64(size) * 0.62)
	return fmt.Sprintf(`<table role="presentation" align="center" cellspacing="0" cellpadding="0" border="0" style="margin:0 auto;border-collapse:separate"><tr><td bgcolor="#10B981" width="%d" height="%d" align="center" valign="middle" style="width:%dpx;height:%dpx;background:linear-gradient(135deg,#34D399 0%%,#10B981 55%%,#047857 100%%);background-color:#10B981;border-radius:%dpx;text-align:center;vertical-align:middle;line-height:0;box-shadow:0 6px 18px -6px rgba(16,185,129,0.55)"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64" width="%d" height="%d" style="display:inline-block;vertical-align:middle" aria-hidden="true"><path d="M32 18 L34 30 L46 32 L34 34 L32 46 L30 34 L18 32 L30 30 Z" fill="#ffffff"/></svg></td></tr></table>`,
		size, size, size, size, radius, starSize, starSize)
}
